hail/src/main/scala/is/hail/expr/ir/Parser.scala

```diff
@@ -766,7 +766,13 @@ object IRParser {
       case "MakeArray" =>
         val typ = opt(it, type_expr(env.typEnv)).map(_.asInstanceOf[TArray]).orNull
         val args = ir_value_children(env)(it)
-        MakeArray.unify(args, typ)
+
+        val finalType = if(typ != null) {
+          typ
+        } else {
+          TArray(args.head.typ)
+        }
+        MakeArray(args, finalType)
```

Comment:
I think this is a more direct, easier to follow construction of MakeArray, compared with MakeArray.unify. It may also generate better IR: the unify path would generate an ArrayMap (via PruneDeadFields.upcast) if args were length > 0. However, the cases that I've seen typ != null are also those in which args are length 0, so that path may have never been triggered.

I believe this change is sound, and less error prone than what we had before: If a typ is provided that attemtpts to change the elementType to something other than args.all, TypeCheck will catch it.

Reply:
I think this is wrong (until we remove requiredness from virtual types). If you have:
```
MakeArray(Seq(In(0, PInt32(true)), In(1, PInt32(false)))`
```

You should definitely not get back a `TArray(TInt32(true))`

Reply:
Shouldn't we be making changes assuming that requiredeness on virtual types is not being used, and check that our code doesn't break/isn't relying on it? Otherwise I imagine the TypeCheck and PruneDeadFields changes to isOfType will have a disruptive effect.

Reply:
no, that makes me extremely nervous. I'm happy with these changes to the If/Coalesce/MakeArray nodes, because we can preserve the 1-1 ptype-vtype relationship while making them more flexible in the children they accept (putting the logic in InferType instead of some unify method)

Reply:
Ok, there is some interdependency here that is complicated. PruneDeadFields.upcast must `assert(ir.typ isOfType rType)` rather than == with the recent Coalesce and If changes iirc, which has broad impact. Similarly the TypeCheck changes, even if just limited to those 3 nodes, are far reaching, because a MakeArray can be a child to some other node that wants all of its children to have equality in virtual type. Am I not thinking about this correctly?

Reply:
Why? Coalesce/If have checks internally that see whether they need to upcast or not, right?

Reply:
> a MakeArray can be a child to some other node that wants all of its children to have equality in virtual type.

I don't know what you mean by this. Typecheck rules operate locally on each node, can you give an example of how MakeArray could be a problem?

Reply:
Only if there is a requestedType explicitly passed to them. They no longer upcast by setting deepOptional one of their children (head), forcing all others to be the same requiredeness.

Reply:
Hmm. I think that the type inference rules for these nodes need to look like:

```
case If(_, cnsq, altr) =>
    if (cnsq.typ == altr.typ)
        cnsq.typ
    else
        cnsq.typ.deepOptional()
```

Reply:
to my above point, we have this rule in PruneDeadFIelds.rebuild:
```scala
      case If(cond, cnsq, alt) =>
        val cond2 = rebuildIR(cond, env, memo)
        val cnsq2 = rebuildIR(cnsq, env, memo)
        val alt2 = rebuildIR(alt, env, memo)

        if (cnsq2.typ.isOfType(alt2.typ) && requestedType.isOfType(cnsq2.typ))
          If(cond2, cnsq2, alt2)
        else
          If(cond2,
            upcast(cnsq2, requestedType),
            upcast(alt2, requestedType)
          )
```

upcast can look at type equality (with requiredness) and this still works fine.

Reply:
oh, oops, we don't need the `requestedType.isOfType` stuff here (same commetn as on your coalesce PR, I just missed it this one). I'll pr a fix.

Reply:
#7977

Reply:
> Hmm. I think that the type inference rules for these nodes need to look like:

```scala
// InferType
case MakeArray(args, t) =>
        if(args.forall(a => a.typ.isOfType(args.head.typ)))
          t.copy(args.head.typ) //keep the outer requiredeness
        else
          t.copy(args.head.typ.deepOptional()) //keep the outer requiredeness?
```

What about the top-level requiredess?

Reply:
I think keeping outer requiredeness as is should be fine; certainly won't affect upcast as called in PruneDeadFields.rebuildIR, nor MakeArray branch in Parser. Inner element requiredeness and wrapping container requiredenss are orthogonal.

Reply:
For coalesce: https://github.com/hail-is/hail/issues/7980

Reply:
the InferType rule for MakeArray shouldn't change -- the `t` should be the type of the node. But wherever MakeArray is constructed, we should be sure that the type we're giving the node is correct.

Reply:
Ah, I misunderstood "Hmm. I think that the type inference rules for these nodes need to look like"

Reply:
yeah, we probably want a helper function for inferring the MakeArray `t` that we call in the pruner and parser.

Reply:
Sorry this wasn't smoother on this PR, tricky at the transition.

Reply:
don't worry, this stuff is hard. I've had more than a year to think through these changes, you've just had a couple of months!

Reply:
Why is a MakeArray with no args valid?

Reply:
empty arrays are totally valid

Reply:
I see, so 1) We don't don't want a special construct in this case to distinguish from a misuse of MakeArary (such as MakeArray(Seq(PVoid), typ) ), and 2) our python-side IR emitter won't elide that if it isn't followed by any uses of the array ... this is probably the wrong thing for python to care about, and we probably remove it in Prune Dead Fields if it isn't used (I could imagine we would rewrite the operation to defer making the array until the first use, rather than make empty and append).

Is this right?

Lastly, why isn't Coalesce allowed to 0 args?

Reply:
I don't quite understand what you mean about dead code elimination in your first point. 

The reason we don't need a 0-arg coalesce is that a no-arg coalesce would just be a missing value, so we can use NA for that.

Reply:
Thanks. Meant that if an empty arrow is created and unused it can be safely eliminated, avoid construction, memory allocation